Description: Revert sage commit 308fa0258773ac7a9b3d797aab268febaa09846e
 This is dependent on a python patch that hasn't landed upstream or in Debian
 yet. This fixes segfaults in the docbuild. TODO: ask Debian Python maintainer
 to apply the patch from #25750 below, and drop this patch.
Bug: https://trac.sagemath.org/ticket/19633
Bug-Python: https://bugs.python.org/issue25750
Author: Ximin Luo <infinity0@debian.org>
Forwarded: not-needed
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/sage/src/sage/misc/lazy_import.pyx
+++ b/sage/src/sage/misc/lazy_import.pyx
@@ -53,7 +53,7 @@
 #*****************************************************************************
 from __future__ import absolute_import
 
-from cpython.object cimport PyObject_RichCompare
+from cpython.object cimport Py_LT, Py_LE, Py_EQ, Py_NE, Py_GT, Py_GE
 
 import os
 from six.moves import cPickle as pickle
@@ -73,7 +73,6 @@
 # boolean to determine whether Sage is still starting up
 cdef bint startup_guard = True
 
-
 cpdef finish_startup():
     """
     This function must be called exactly once at the end of the Sage
@@ -175,7 +174,9 @@
         self._at_startup = at_startup
         self._deprecation = deprecation
 
-    cpdef _get_object(self, owner=None):
+    # Due to a bug in Cython-0.19, this must not be a cpdef method.
+    # See http://trac.sagemath.org/sage_trac/ticket/14452
+    def _get_object(self, owner=None):
         """
         Return the wrapped object, importing it if necessary.
 
@@ -234,41 +235,39 @@
                sage: type(Foo.__dict__['plot'])
                <... 'function'>
         """
-        if self._object is not None:
-            return self._object
-
-        if startup_guard and not self._at_startup:
-            import sys, traceback
-            print('-' * 79)
-            print('Resolving lazy import {0} during startup'.format(self._name))
-            print('Calling stack:')
-            traceback.print_stack(None, None, sys.stdout)
-            print('-' * 79)
-        elif self._at_startup and not startup_guard:
-            print('Option ``at_startup=True`` for lazy import {0} not needed anymore'.format(self._name))
-        self._object = getattr(__import__(self._module, {}, {}, [self._name]), self._name)
-        alias = self._as_name or self._name
-        if self._deprecation is not None:
-            from sage.misc.superseded import deprecation
-            try:
-                trac_number, message = self._deprecation
-            except TypeError:
-                trac_number = self._deprecation
-                message = None
-            if message is None:
-                message = ('\nImporting {name} from here is deprecated. ' +
-                    'If you need to use it, please import it directly from' +
-                    ' {module_name}').format(name=alias, module_name=self._module)
-            deprecation(trac_number, message)
-        if owner is None:
-            if self._namespace and self._namespace[alias] is self:
-                self._namespace[alias] = self._object
-        else:
-            from inspect import getmro
-            for cls in getmro(owner):
-                if cls.__dict__.get(alias, None) is self:
-                    setattr(cls, alias, self._object)
-                    break
+        if self._object is None:
+            if startup_guard and not self._at_startup:
+                import sys, traceback
+                print '-' * 79
+                print 'Resolving lazy import {0} during startup'.format(self._name)
+                print 'Calling stack:'
+                traceback.print_stack(None, None, sys.stdout)
+                print '-' * 79
+            elif self._at_startup and not startup_guard:
+                print 'Option ``at_startup=True`` for lazy import {0} not needed anymore'.format(self._name)
+            self._object = getattr(__import__(self._module, {}, {}, [self._name]), self._name)
+            alias = self._as_name or self._name
+            if self._deprecation is not None:
+                from sage.misc.superseded import deprecation
+                try:
+                    trac_number, message = self._deprecation
+                except TypeError:
+                    trac_number = self._deprecation
+                    message = None
+                if message is None:
+                    message = ('\nImporting {name} from here is deprecated. ' + 
+                        'If you need to use it, please import it directly from' + 
+                        ' {module_name}').format(name=alias, module_name=self._module)
+                deprecation(trac_number, message)
+            if owner is None:
+                if self._namespace and self._namespace[alias] is self:
+                    self._namespace[alias] = self._object
+            else:
+                from inspect import getmro
+                for cls in getmro(owner):
+                    if cls.__dict__.get(alias, None) is self:
+                        setattr(cls, alias, self._object)
+                        break
         return self._object
 
     def _get_deprecation_ticket(self):
@@ -471,7 +470,18 @@
             left = (<LazyImport>left)._get_object()
         if isinstance(right, LazyImport):
             right = (<LazyImport>right)._get_object()
-        return PyObject_RichCompare(left, right, op)
+        if op == Py_LT:
+            return left <  right
+        elif op == Py_LE:
+            return left <= right
+        elif op == Py_EQ:
+            return left == right
+        elif op == Py_NE:
+            return left != right
+        elif op == Py_GT:
+            return left >  right
+        elif op == Py_GE:
+            return left >= right
 
     def __len__(self):
         """
@@ -517,7 +527,7 @@
         documentation of :meth:`_get_object` for an explanation of
         this.
         """
-        obj = self._get_object(owner)
+        obj = self._get_object(owner=owner)
         if hasattr(obj, "__get__"):
             return obj.__get__(instance, owner)
         return obj
